\documentclass[10pt,a4paper]{article}
\usepackage[utf8]{inputenc}
\usepackage[obeyDraft]{todonotes}
\usepackage{hyperref}
\usepackage{amssymb}
\usepackage{longtable}
\usepackage{multirow}
\usepackage[T1]{fontenc}
% --------------------------------------------------------------------------------------------------
\title{JACUSA2 manual}
\author{Michael Piechotta \\ michael.piechotta@gmail.com}
\date{8th, October, 2019} 
% --------------------------------------------------------------------------------------------------
\begin{document}
\newcommand{\call}[1]{{\textit call-#1}}
\newcommand{\pileup}{{\textit pileup}}
\newcommand{\rtarrest}{{\textit rt-arrest}}
\newcommand{\lrtarrest}{{\textit lrt-arrest}}
% --------------------------------------------------------------------------------------------------
\maketitle 
\tableofcontents
\listoftodos
%---------------------------------------------------------------------------------------------------
\section{Introduction}
JAVA framework for accurate Variant assessment (JACUSA2) is a one-stop solution to detect single
nucleotide variants (SNVs) and reverse transcriptase induced arrest events in Next-generation 
sequencing (NGS) data.

\url{https://github.com/dieterich-lab/JACUSA2/}{JACUSA2} is a direct successor of 
\url{https://github.com/dieterich-lab/JACUSA/}{JACUSA1} --- JACUSA1 is hereby deprecated and won't be 
continued. All methods from JACUSA1 (\call{1}, \call{2}, and \pileup) are available in JACUSA2.
The new release of JACUSA2 features great performance enhancements (~3 faster) for existing methods 
and adds new methods (\rtarrest and \lrtarrest) that enable to identify read arrest events.
\begin{figure}[ht]
	\centering
	\includegraphics[width=\textwidth]{figures/poster_jacusa_methods}
	\caption{Schematic summarising JACUSA2 methods and underlying data}
	\label{fig:methods}
\end{figure}
\rtarrest allows to identify read arrest events by comparing read through and read arrest counts. 
\lrtarrest is a combination of variant and arrest event detection that allows to identify linked events.

Some artefact filters from JACUSA have been moved to the R helper package called 
\url{https://github.com/dieterich-lab/JACUSAhelper2}{JACUSA2helper}.
A new artefact filter has been added to JACUSA2 that allows to filter candidate variants or arrest events 
by providing a file that contains sites to be excluded.

Another new feature in JACUSA2, are optional data that can be added to the output.
Such optional data are INDEL counts and associated differential statistics and read substitution information.
A user defined base substitution can be used to partition reads into two sets: reads containing the 
user specified base substitution and total reads.

JACUSA2 employs a window-based approach to traverse provided BAM\cite{HengLi2009} files featuring highly 
parallel processing and utilizing the \url{https://github.com/samtools/htsjdk}{htsjdk} framework.
% --------------------------------------------------------------------------------------------------
\subsection{Variant calling}
Robust identification of variants has proven to be a daunting task due to artefacts specific for 
NGS-data and employed mapping strategies. We implement various artefact filters that reduce the number 
of false positives (see section \ref{sec:filters}).

JACUSA2 has been extensively evaluated and optimized to identify RNA editing sites in RNA-DNA and
RNA-RNA sequencing samples. Checkout the original publication and the supplementary material of 
JACUSA1 \cite{Piechotta2017} if you are interested in details regarding the test-statistic.
% --------------------------------------------------------------------------------------------------
\subsection{Reverse transcriptase arrest events}
\begin{figure}[ht]
  \centering
  \includegraphcs[width=\textwidth]{figures/arrest\_events}
  \caption{Schematic depiction of arrest events that have been induced by modifications that influence 
  reverse transcription during library preparation}
  \label{fig:arrest_events}
\end{figure}
Reverse transcriptase arrest events can be induced during library preparation (see Figure. \ref{fig:arrest_events}).
They are identified by reads that exhibit shorter than expected read length due to premature termination 
during first strand synthesis. For each site, a vector of read through and read arrest counts is constructed 
and modelled with a Beta-Binomial distribution. We estimate the parameters of the distribution with 
method presented by Minka\footnote{See \url{https://tminka.github.io/papers/dirichlet/minka-dirichlet.pdf}{Minka-Dirichlet}}.
In order to represent arrest events in JACUAS2 output, the base vector colum ``basesij''  is split into 
two columns ``arrest\_basesij'' and ``through\_basesij''. The former column represents base calls from 
reads that exhibit premature termination and the later corresponds to base calls from reads that 
show NO premature termination.
% --------------------------------------------------------------------------------------------------
\section{Download}
The latest version of JACUSA2 can be obtained from \url{https://github.com/dieterich-lab/JACUSA2/}{JACUSA2}.
You can go to \url{https://github.com/dieterich-lab/JACUSA2/releases}{all releases} and pick the lastest release.
% --------------------------------------------------------------------------------------------------
\subsection{Installation and requirements}
JACUSA2 does not need any configuration but needs a correctly configured Java environment.
We developed and tested JACUSA2 with Java v1.8. If you encounter any Java related problems please
consider to change to Java v1.8.
% --------------------------------------------------------------------------------------------------
\subsection{Migrating from JACUSA1 to JACUSA2}\label{sec:migration}
There are several important changes to the command line interface:
\begin{itemize}
  \item ONLY single dash ``-'' options, e.g.: ``-c 10'' are supported. ALL two dash options ``--option [\ldots]'' from JACUSA1 have been removed.
  \item Use ``\textbf{-}filterNH'' and ``\textbf{-}filterNM'' instead of ``--filterNH'' and ``--filterNM''.
	\item CLI option to provide library type has changed: JACUSA1: ``-P Lib1,Lib2'', JACUSA2: ``-P1 Lib1 -P2 Lib2''.
\end{itemize}
In JACUSA2 a \#\#'' prefixed header line that contains command line options and used version is added 
to the default output file format. There is also a new version of \url{https://github.com/dieterich-lab/JACUSA2helper/}{JACUSA2helper}
to support analysis of JACUSA2 output.
% --------------------------------------------------------------------------------------------------
\subsection{Sample \textit{in silico} data}
\subsubsection{Variant calling}
You can choose between different setups to detect variants. The gDNA vs. cDNA represents the typical 
data setup that is encountered in detection of RNA editing sites via comparing genomic and transcriptomic 
sequencing reads. In this setup, variants have been only imputed to the cDNA BAM file. 
The cDNA vs. cDNA data setup can be interpreted as representing allele specific expression of single 
variants or differential RNA editing. In this setup, variants with pairwise different base frequency 
have been imputed into both cDNA BAM files. Additionally, to make the identification of variants more 
challenging SNPs with pairwise similar base frequencies have been included to both BAM files. 
SNPs should not be identified as true positive sites.

gDNA data has been simulated with
art\footnote{\url{http://www.niehs.nih.gov/research/resources/software/biostatistics/art/}{art}}
and cDNA reads have been simulated with
flux\footnote{\url{http://sammeth.net/confluence/display/SIM/Home}{flux simulator}}. Read
simulations have been restricted to the corresponding first chromosome of human.
The archives have the following data:
\begin{description}
  \item[*.bam] BAM files: gDNA.bam and cDNA OR cDNA\_1.bam and cDNA\_2.bam
  \item[variants.txt] Coordinates of imputed variants and their target and sampled frequencies
  \item[(snps.txt)] Only available for cDNA vs. cDNA. Coordinates of imputed SNPs. In both
  BAM files matching SNPs have the same target frequency but different effective or sampled
  frequency. The shape parameter determines how much the sampled frequency will deviate from the
  target frequency in each BAM file. The suffixes: \_cdna\_1 and \_cdna\_2 correspond to the
  respective BAM file]
\end{description}
Available sample data:
\begin{itemize}
  \item
  \item \url{https://data.dieterichlab.org/s/hg19_chr1_gDNA_VS_cDNA}{hg19\_chr1\_gDNA\_VS\_cDNA}
  \item \url{https://data.dieterichlab.org/s/hg19_chr1_cDNA_VS_cDNA}{hg19\_chr1\_cDNA\_VS\_cDNA} TODO
  %\item
  %\url{http://www.age.mpg.com/software/jacusa/sample_data/hg19_chr1_cDNA_VS_cDNA.tar.gz}{hg19\_chr1\_cDNA\_VS\_cDNA.tar.gz}
\end{itemize}
% --------------------------------------------------------------------------------------------------
\subsection{Reverse transcriptase arrest event}
We have downloaded primary sequencing data from \cite{Zhou2018} and mapped them according to \todo[author=christoph]{How did we do this?}.
From the resulting set, we retained only uniquely mapping reads with mapping quality $\ge 20$.
\url{https://data.dieterichlab.org/s/data_Zhou2018}{data\_Zhou2018} \todo[author=michael]{Upload data and link?}
\cite{Zhou2018} map transcriptome-wide RNA modification of pseudouridine (\PSI) by chemically modifying 
pseudouridines with carbodiimide (+CMC) and detecting arrest events that were induced by reverse transcription 
stops in high-throughput sequencing. In summary, +CMC and mock-treated HEK293T rRNA was reverse 
transcribed under 3 different conditions (HIVRT, SIIIMn, and SIIIMg) and seqeunced.

The archive \url{https://data.dieterichlab.org/s/data_Zhou2018}{data\_Zhou2018} consists of:
\begin{description}
  \item[*.bam, *.bam.bai] BAM files and indicices for different 3 conditions HIVRT, SIIIRTMg, and SIIIRTMn each treated 
  with CMC(+CMC) or mock-treated(-CMC) (6 BAM + 6 BAI files).
  \item[fournier\_db.txt] Parsed and coordinate adjusted list of known modifications for human 18S and 28S 
  according to Fournier lab's 3D rRNA modification maps database \cite{PieknaPrzybylska2007}. 
\end{description}
%---------------------------------------------------------------------------------------------------
\section{Input}
All JACUSA2 methods require sorted and indexed \url{https://samtools.github.io/hts-specs/SAMv1.pdf}{BAM} files.
BAM is a standardized file format for efficient storage of alignments.
Furthermore, JACUSA2 requires that the reference sequence is available either through the ``MD'' 
\url{https://samtools.github.io/hts-specs/SAMtags.pdf}{tag} in BAM files or by providing the reference 
sequence in indexed FASTA format with the command line option ``-R <reference.fasta>''.
The ``MD'' field contains mismatch information that allows to perform variant calling without providing 
the reference sequence.

Check the manuals of: \url{http://samtools.sourceforge.net/}{SAMtools/BCFtools} and/or
\url{http://broadinstitute.github.io/picard/}{picard tools} for how to use the
respective tool to convert your alignment files to valid JACUSA2 input BAM.
%---------------------------------------------------------------------------------------------------
\subsection{Processing BAM files}
In the following, commands for SAMtools are presented.

To sort and index your raw BAM files perform the following sequence of commands:
\begin{description}
\item[SAM $\rightarrow BAM$] \begin{verbatim} samtools view -Sb mapping.sam > mapping.bam \end{verbatim}
\item[sort BAM] \begin{verbatim} samtools sort mapping.bam mapping.sorted \end{verbatim} 
\item[index BAM] \begin{verbatim} samtools index mapping.sorted.bam \end{verbatim}
\end{description}

Check your BAM file for the ``MD'' \url{https://samtools.github.io/hts-specs/SAMtags.pdf}{tag} 
if you want to provide reference sequence information via this tag. 
When your BAM files do not have the ``MD'' tag set correctly use SAMtools:
\begin{verbatim}
samtools calmd mapping.sorted.bam reference.fasta > mapping.sorted.MD.bam
\end{verbatim}
%---------------------------------------------------------------------------------------------------
\subsubsection{Remove duplicates for variant calling}
It is a recommended pre-processing step to remove duplicate reads when identifying variants - \textbf{omit 
this step for ``rt-arrest'' and ``lrt-arrest''}. Reads that are terminated prematurely during 
library preparation will falsely be identified as PCR duplicates and removed from the final output. 
This will dramatically degrade sensitivity of arrest event identification.

Duplicated reads occur mostly due to PCR-artefacts. They are likely to harbour false variants and 
most statistical test require that reads are sampled independently.
In the following, commands for picard tools are presented:
\begin{verbatim} java -jar MarkDuplicates.jar \ 
  I=mapping.sorted.bam O=dedup_mapping.sorted.bam \ 
  M=duplication.info
\end{verbatim}
Most tools that filter duplicates reads either remove duplicated reads from the final output or assign 1024 
to the flag field of those reads. In the later case, invoke JACUSA2 with the additional command line 
option ``-F 1024'' to filter reads that have been marked as duplicates.
%---------------------------------------------------------------------------------------------------
\subsubsection{Library type and strand information}
JACUSA2 supports stranded paired end and single ends reads. Warning: the CLI option has changed 
(see Section \ref{sec:migration}).
With the command line parameter ``-P <LIBRARY-TYPE> | -P1 <LIBRARY-TYPE> -P2 <LIBRARY-TYPE>'' the 
user can choose the underlying library type:
\begin{description} 
\item[RF-FIRSTSTRAND] STRANDED library - first strand sequenced,
\item[FR-SECONDSTRAND] STRANDED library - second strand sequenced, and
\item[UNSTRANDED] UNSTRANDED library.
\end{description}
The UNSTRANDED library type is not available for rt/lrt-arrest method because an arrest site can not 
unambiguously be defined for this library type. Read base substitution option ``-B <BASE-SUB>'' 
requires a stranded library type in order to unambiguously identify base substitutions.
%---------------------------------------------------------------------------------------------------
\subsection{Traverse BED-like file}
Identification of interesting sites can be restricted to specific regions of the genome or transcriptome. 
Provide a minimalistic BED-like file to limit the search to this region(s) or site(s). 
Remaining region(s) of the BAM files will not be considered.

In the following traverse file, the search is confined to a 100nt region on contig 1
starting at position 1,000 and a single site on contig 2 at coordinates 10,000:
\begin{table}
\centering
\caption{Example of BED-like traverse file}
\label{tb:traverse_file}
\begin{tabular}{lll}
\textbf{contig} & \textbf{start} & \textbf{end} \\
\hline
1 & 1000 & 1100 \\
2 & 10000 & 10000 \\
\multicolumn{3}{c}{}
\end{tabular}
\end{table}
Many individual sites may impair running performance of JACUSA2.
If possible, try to merge nearby sites into contiguous regions and extract specific sites from 
JACUSA2 output with \url{http://bedtools.readthedocs.org/en/latest/}{bedtools} ``intersect'':
\begin{description}
\item[merge sites] \begin{verbatim} 
bedtools merge -d 500 singular_sites.bed > \ 
  contigous_regions.bed
\end{verbatim}
\item[run JACUSA2] \begin{verbatim} 
java -jar JACUSA2.jar call-2 -b contigous_regions.bed -r
JACUSA2.out mapping_1.sorted.bam mapping_2.sorted.bam
\end{verbatim}
\item[extract sites] \begin{verbatim}
bedtools intersect -wa -a JACUSA2.out -b singular_sites.bed
\end{verbatim}
\end{description}
%---------------------------------------------------------------------------------------------------
\subsection{Output}
JACUSA2 writes its output to a user speficied file. When using multiple threads, JACUSA2 will
create a temporary file for each allocated thread in the temporary directory that is provided by the 
JAVA Virtual Machine. Chosen command line parameters and current genomic position are printed to the 
command prompt and serve as a status guard. Furthermore, depending on the provided command line 
parameters, JACUSA2 will generate a file with sites that have been identified as potential artefacts 
when ``-s'' is provided. Currently, JACUSA2 supports the following output formats, controlled by ``-f'':
\begin{itemize}
  \item Default (JACUSA2 output --- varies between JACUSA2 methods)
  \item Variant Call Format (VCF)\footnote{\url{http://samtools.github.io/hts-specs/VCFv4.1.pdf]}{VCF file format}}
\end{itemize}
The default output format is based on
BED6\footnote{\url{http://genome.ucsc.edu/FAQ/FAQformat.html\#format1}{BED file format}} with
additional JACUSA2 methods specific columns. The actual number of columns depends on the JACUSA2 
method and the number of provided BAM files.
\begin{table}[ht]
\caption{JACUSA2 default output format --- core elements}
{\small
\begin{tabular}{lcccccc|ccc}
Column: & 1 & 2 & 3 & 4 & 5 & 6 & \ldots & N-1 & N \\
\hline
& 1 & 100 & 101 & variant & $8.07\ldots$ & - & JACUSA2 method specific & * & * \\	
\multicolumn{6}{c|}{\ldots} & \multicolumn{4}{c}{\ldots}
\end{tabular}}
\end{table}
\begin{description}
\item[(1, 2, 3) contig + start + end] 0-based, genomic coordinates.
\item[(4) name] Currently, constant string that depends on each method. This dummy field is to ensure 
BED6 compatibility.
\item[(5) score] Test-statistic $z \in \mathbb{R}$ that indicates the likelihood that this is a true
variant. Higher number indicates a higher likelihood for a variant.
\item[(6) strand] Possible values are: ``.'', ``+'', and ``-'' which correspond to ``unstranded'',
``positive strand'', and ``negative strand'', respectively. If strand is != ``.'', then the following b
ase columns will be indicating base counts according to the strand - inverted base count if on the 
``negative strand''.
\item[(7-N-2) method specific] The number of base columns depends on the JACUSA2 method --- check method specific explanation.
\item[(N-1) info] Additional info for this specific site. Currently, details about the parameter
estimation of the underlying distribution can be shown, insertion/deletion counts and statistics, 
and additional method specific data. If nothing provided, the empty field is equal to ``*''
\item[(N) filter\_info] Relevant, if feature filter(s) $X$ have been provided with ``-a X'' on the
command line. The column will contain a comma-separated list of feature filters that predict this
site to be a potential artefact. Possible values depend on the utilized JACUSA2-method \\ 
\end{description}
%---------------------------------------------------------------------------------------------------
\section{Feature/Artefact filter}
\begin{figure}[h]
\centering
\includegraphics[width=\textwidth]{figures/jacusa\_filter\_cropped}
\end{figure}
\begin{tabular}{lp{.8\textwidth}}
\textbf{Value} & \textbf{Description of potential artefact} \\
\hline
D & Variant call in the vicinity of Read Start/End, Intron, and/or INDEL position \\
B & Variant call in the vicinity of Read Start/End \\
I & Variant call in the vicinity of INDEL position \\
S & Variant call in the vicinity of Splice Site \\
Y & Variant call in the vicinity of homopolymer \\
M & Max allowed alleles exceeded \\ 
H & ``Control'' sample contains non-homozygous pileup \\
\end{tabular}
%---------------------------------------------------------------------------------------------------
\section{Variant detection}
\subsection{Identification of RNA editing sites}
In order to identify RNA editing sites by comparing gDNA and \emph{stranded} RNA-Seq (single or paired end) use:
\begin{description} 
\item[first strand sequenced] ``-P1 UNSTRANDED -P2 RF-FIRSTSTRAND''
\item[second strand sequenced] ``-P2 UNSTRANDED -P2 FR-SECONDSTRAND''
\end{description}.
When your RNA-Seq is unstranded use: ``-P1 UNSTRANDED -P2 UNSTRANDED'' and infer the correct orientation from annnotation.

Use the following command line to identify RNA-DNA differences in BAM files that might give rise to RNA editing sites:
\begin{verbatim}
java -jar call-2 -r JACUSA.out -s -a H:1 gDNA.bam cDNA.bam
\end{verbatim}
Option ``-a H:1'' ensures that potential polymorphisms in gDNA will be eliminated as artefacts. The number $x \in \{1, 2\}$
determines which sample has to be homomorph - in this case: gDNA.bam.

Use the following command line to identify RNA-DNA differences:
\begin{verbatim}
java -jar call-2 -r JACUSA2.out -s cDNA1.bam cDNA2.bam
\end{verbatim}
WARNING: If you want to identify RNA-RNA differences make sure NOT to use the filter ``-a H:x''! Otherwise, potential valid variants will be filtered out. 
%---------------------------------------------------------------------------------------------------
\section{Reverse transcriptase arrest events}
%\todo[author=michael]{add text}
\section{Usage}
Calling JACUSA2 without any arguments will print the available tools which currently are:
\begin{verbatim}
java -jar JACUSA2.jar
  METHOD        DESCRIPTION
  call-1        Call variants - 1 condition
  call-2        Call variants - 2 conditions
  pileup        SAMtools like mpileup (2 conditions)
  rt-arrest     Reverse Transcription Arrest - 2 conditions
  lrt-arrest    Linkage arrest to base substitution - 2 conditions
Version: 		[...]
Libraries: 	
\end{verbatim}
%---------------------------------------------------------------------------------------------------
\subsection{call-1}
Single sample (call-1) allows to call variants against a reference. 
Internally, an \textit{in silico} sample is created from information that is provided by the ``MD'' field 
in BAM files.

The number of base columns depends on the number of BAM files. In basesIJ: $I$
corresponds to sample and $J$ to the respective replicate. Numbers indicate the base count of the
following base vector: $(A, C, G, T)$

Sites that have a $>$ alleles are considered candidate variant sites and for this sites a test-score will be computed.
\subsection{call-2}
%\todo[author=michael]{nicely combine call-1, call2}
%---------------------------------------------------------------------------------------------------
\subsection{pileup}
See ``Call variant - two samples'' for details.
\subsection{rt-arrest - 2 conditions}
In this method base call counts of arrest and read through reads are modelled by a Beta-Binomial distribution and 
differences between conditions are to be identified by means of a likelihood-ratio test. Subsequent approximiation 
with $\chi^2$ distribution to compute a pvalue.

Sites are considered candidate arrest sites, if in all BAM files there is at least one read through AND one  
read arrest event. Furthermore, coverage filter and minBASQ of Base Call apply that will affect the output. 
\subsection{lrt-arrest - 2 conditions}
%\todo[author=michael]{make fluent text}
lrt-arrest allows to link pileups to their arrest position. Output consists of read arrest and read through counts and 
a references to the associated arrest positions. There are cases, where currently an arrest position cannot be defined, 
e.g.: non properly paired reads.
Output consits of at least one line. Each separate arrest position adds an additional row is 
The first row contains the unstratified data or total, the ``arrest\_pos'' column is set to ``*''.
Any following sites with identical coordinates (contig, start, end, strand) will have a different 
arrest position reference in the ``arrest\_pos'' column. 

This method supports partial artefact filtering. Currently, filters only apply to the unstratified data --- 
sites with ``*'' in in ``arrest\_pos''. Furthermore, coverage filter and minBASQ of Base Call apply 
that will affect the output.
%---------------------------------------------------------------------------------------------------
% TODO options
\section{Description of command line options}
\subsection{Input / Output}
\subsubsection{Input BAM files}
% TODO BAM files
\subsubsection{Output file}
\input{CLI/r}
%
\subsection{Filtering artefacts}
\subsubsection{Configure feature filter}
\input{CLI/a}
\subsubsection{Output artefacts to separate file}
\input{CLI/s}
%
\subsection{Input BED file}
\input{CLI/b}
\subsection{Reference fasta file}
\input{CLI/R}
%
\subsection{Library type}
\input{CLI/P}
%
\subsection{Read base changes}
\input{CLI/B}
\subsection{Show deletion score}
\input{CLI/D}
\subsection{General filtering}
\subsubsection{Filter by mapping quality}
\input{CLI/m1}
\subsubsection{Filter by base call quality}
\input{CLI/q1}
\subsubsection{Filter by minimal coverage}
\input{CLI/c1}
\subsubsection{Limit maximal depth}
\input{CLI/d1}
\subsection{Specific filtering}
\subsubsection{Filter by flag(s)}
% TODO escape latex letters \input{CLI/f}
\subsubsection{Retain by flag(s)}
\input{CLI/F}
\subsubsection{Filter by number of hits}
\input{CLI/filterNH_1}
\subsubsection{Filter by number of mismatches}
\input{CLI/filterNM_1}
% 
\subsection{Thread related}
\subsubsection{Number of parallel threads}
\input{CLI/p}
\subsubsection{Actual thread window size}
\input{CLI/w}
\subsubsection{Reserved thread window size}
\input{CLI/W}
%
\subsection{Test-statistic options}
\input{CLI/T}
\input{CLI/u}
\subsection{Filtering by Test-statistic threshold}
% TODO remove
%\clioption{-T <THRESHOLD>}{Filter positions based on test-statistic THRESHOLD\newline default: DO 
%NOT FILTER}{\callone \calltwo \rtarrest \lrtarrest}
\subsection{Misc}
\input{CLI/h}
\input{CLI/x}
%---------------------------------------------------------------------------------------------------
\section{Used libraries}
\begin{tabular}{lrl}
\bf{Libray} & \bf{Version} & \bf{Source} \\
\hline
htsjdk & 2.12.0 & \url{https://github.com/samtools/htsjdk} \\
Apache commons-cli & 1.4 & \url{https://commons.apache.org/proper/commons-cli} \\
Apache commons-math3 & 3.6.1 & \url{http://commons.apache.org/proper/commons-math}
\end{tabular}
%---------------------------------------------------------------------------------------------------
\bibliographystyle{apalike}
\bibliography{manual}
%---------------------------------------------------------------------------------------------------
\end{document}